util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java

```diff
@@ -0,0 +1,204 @@
+                                                    () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MAY),
+                                                    () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MUST));
+
+        CompactMTS<String> context = MTSUtil.conjunction(greenContext, redContext);
+
+        Assert.assertTrue(MTSUtil.isRefinementOf(context, instance.context, context.getInputAlphabet()));
+    }
+
+    @Test(dataProvider = "Decomp")
+    public void testDecompContextReverse(DecompositionTest testCase) throws IOException {
+        if (testCase.modal_contract == null) {
+            return;
+        }
+        testCase.context = "src/test/resources/phil3/" + testCase.context;
+        testCase.modal_contract = "src/test/resources/phil3/" + testCase.modal_contract;
+        testCase.system = "src/test/resources/phil3/" + testCase.system;
+        testCase.orig_sys = "src/test/resources/phil3/" + testCase.orig_sys;
+        DecompositionInstance instance = new DecompositionInstance(testCase);
+
+        LOGGER.debug("Testing decomp for "+testCase.orig_sys);
+        LOGGER.debug("Modal contract "+testCase.modal_contract);
+        LOGGER.debug("Components: "+testCase.context + ", " +testCase.system);
+        LOGGER.debug("Com. alphabet: {}", instance.modal_contract.getCommunicationAlphabet());
+
+        MCUtil.SystemComponent<CompactMTS<String>, Integer, String, MTSTransition<String, MutableModalEdgeProperty>, MutableModalEdgeProperty> systemComponent = MCUtil.systemComponent(instance.modal_contract,
+                new CompactMTS.Creator<String>(),
+                (x) -> new ModalEdgePropertyImpl(x.getProperty().getType()),
+                () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MAY));
+
+        CompactDFA<String> redLanguage = (CompactDFA<String>) MCUtil.redContextLanguage(systemComponent, instance.modal_contract.getCommunicationAlphabet());
+        CompactMTS<String> redContext = MCUtil.redContextComponent(redLanguage,
+                new CompactMTS.Creator<String>(),
+                instance.modal_contract.getCommunicationAlphabet(),
+                () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MAY));
+
+        CompactDFA<String> greenLanguage = (CompactDFA<String>) MCUtil.greenContextLanguage(instance.modal_contract);
+        CompactMTS<String> greenContext = MCUtil.greenContextComponent(greenLanguage,
+                new CompactMTS.Creator<String>(),
+                instance.modal_contract.getCommunicationAlphabet(),
+                () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MAY),
+                () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MUST));
+
+        CompactMTS<String> context = MTSUtil.conjunction(greenContext, redContext);
+
+        MTSUtil.saveMTSToPath(context, "src/test/resources/phil3-actual/"+ testCase.context);
```

Comment:
Do you really want to write a file to a source directory? This makes repeated builds (even with `clean`) not produce the same artifacts. Is there any specific reason you need to write this (or any) file? I haven't seen it being read anywhere.

Reply:
You are right, the source directory was a bad choice. It should have been a separate resource directory. The idea is to log (intermediate) graphs since the assertions currently do not provide sufficient error information on failure.

Reply:
I'd argue that debugging is mostly done via an IDE that allows you to directly inspect in-memory objects rather than `System.out.println`-ing intermediate results. Would `LOGGER.debug()` be a viable output alternative? This would not require any file magic and you could still manually copy the terminal output to a file if needed.

Reply:
I like specific tests with descriptive error information. Unfortunately, there is no support for such `assert`s on graphs. Therefore we either need to define our own asserts (and I have done that to some degree in the client code), or we need to compromise.

Dumping the graphs as dot files on the file system was never meant as a long term solution. Nevertheless, it proved really useful. Using a logger would be a cleaner solution, but it is also less practical.

Reply:
I think the best course would be to implement custom `assert`s and then using only logging and no file-utils.

Reply:
I just realized this is test code and not library code :flushed:. I'm fine with having some intermediate checks/asserts as you suggested. Since most serialization code only requires some form of `OutputStream` or `Writer` we could use something like a `StringWriter` which stores the output in an in-memory `String` for future comparisons. This would be more convenient than a logger, but we wouldn't require filesystem access.

However, in this concrete case, you check the serialization of a modal conjunction, which mixes two aspects: The conjunction operation and the serialization of the result. From a unit-testing perspective, checking only the former would better fit the separation of concerns. You already do this with the assertion in the next line. Plus: I haven't seen code that actually reads the `phil3-actual` files anywhere, so you are currently not verifying the output anyway, right?

For now, I would still vote for removing this line, because it currently doesn't serve any purpose. For future (post-reintegrate) steps, we can add more asserts, possibly including in-memory serialization, ok?

Reply:
I like your proposition with the `StringWriter`. Could one automatically flush them in case of an failure?

As for the serialization: it is used to allow debugging. The assert fails with a message like "expected false to be true", which does not help finding the cause of an error.